format/spec.md

```diff
@@ -454,7 +454,7 @@ Partition field IDs must be reused if an existing partition spec contains an equ
 | **`truncate[W]`** | Value truncated to width `W` (see below)                     | `int`, `long`, `decimal`, `string`, `binary`                                                              | Source type |
 | **`year`**        | Extract a date or timestamp year, as years from 1970         | `date`, `timestamp`, `timestamptz`, `timestamp_ns`, `timestamptz_ns`                                      | `int`       |
 | **`month`**       | Extract a date or timestamp month, as months from 1970-01-01 | `date`, `timestamp`, `timestamptz`, `timestamp_ns`, `timestamptz_ns`                                      | `int`       |
-| **`day`**         | Extract a date or timestamp day, as days from 1970-01-01     | `date`, `timestamp`, `timestamptz`, `timestamp_ns`, `timestamptz_ns`                                      | `int`       |
+| **`day`**         | Extract a date or timestamp day, as days from 1970-01-01     | `date`, `timestamp`, `timestamptz`, `timestamp_ns`, `timestamptz_ns`                                      | `int` (the physical type should be an `int`, but the the logical type should be a `date`) |
```

Comment:
Well, it's unclear what is **physical** type versus **logical** type.

Reply:
Do you have better suggestions?

Reply:
How about using using `date` directly here. In the context of iceberg, `date` is primitive type. ref: https://iceberg.apache.org/spec/#nested-types:~:text=38%20or%20less-,date,-Calendar%20date%20without cc @Fokko

Reply:
TBH I also don't fully understand why we can't use `date` directly here, since it's a primitive type.

Reply:
I needed some time to think about this. I would also lean towards stating that it returns a `date`. The confusion (and also the terminology) comes from that it is being encoded in Avro where the physical (bytes written to disk) is an `int`, that's annotated with a logical type that it represents a `date`.

The relevant piece of code:

https://github.com/apache/iceberg/blob/fe2f593cd025223e4ab5ab41a296fb106ce3b1cf/api/src/main/java/org/apache/iceberg/transforms/Dates.java#L92-L98

Curious to learn what others think @RussellSpitzer @rdblue @danielcweeks @nastra

Reply:
There was a discussion on the dev list https://lists.apache.org/thread/2gq7b54nvc9q6f1j08l9lnzgm5onkmx5

Reply:
> TBH I also don't fully understand why we can't use date directly here, since it's a primitive type.

+1. I'm not sure whether my understanding is correct. According to https://github.com/apache/iceberg/issues/279#issuecomment-519620975, the day transform returns an integer from 1970, which is equal to the date representation in iceberg-java so that it can be treated as the date type.

However, in iceberg spec, the definition of date is `Calendar date without timezone or time`, I think this definition is not necessarily equal to `date is an integer from 1970`, and that's why we can't say it's a date in iceberg spec, e.g. in some iceberg implementation, they use string to represent a date. I guess that's what https://github.com/apache/iceberg/pull/9345#discussion_r1439682666 means.

If the above understanding is correct, to make things clear, maybe we should:
1. Change the definition of date to `date is an integer from 1970`. AFAIK, this also holds in iceberg-rust
2. Should not use date as result type, because that's not equal to the definition of date type in spec.

I think using the logical type to explanation is also confused because the spec didn't define what's logical type.

Reply:
I think many previous questions and discussions ended prematurely with "date is also just an integer".  Or "physically as int", but displayed as "date", this is also very confusing.

- Then why not just use date? What may break?
- What on earth is date?
    + In Avro/Parquet, date is exactly a "logical type"
    + In Iceberg spec, it seems not clear what date is physically, and doesn't require it to be an int.

Reply:
I think the spec is clear here for `day` transform. `days from 1970-01-01` is a `int`. What might be confusing is with `date` and implementation details, which can be enhanced like https://github.com/apache/iceberg-python/pull/1211

Reply:
> I think the spec is clear here for day transform. days from 1970-01-01 is a int.

I believe this is not clear enough, and has lead to problems repeately in the wild like https://github.com/apache/iceberg-rust/issues/478.

As also mentioned by Fokko, what is now persisted is really an "Avro Date". Parse it by assuming it's an Avro Int will lead to error.

> When it inserts data, the reference Java Iceberg implementation writes the Avro manifest files, using an Avro type of Date for the partition struct value.

---

Actually this looks a case of abstraction leak to me: We didn't specify `date` is `int` (`days from 1970-01-01`). 

But the `day` transform here requires:
1. The value is `int` (`days from 1970-01-01`)
2. The value should be serialized/displayed as `Date`

This implicitly forces `date` to be `int`. (And then `day` transform's return should also be `date`)

Reply:
> The value should be serialized/displayed as Date (This is not mentioned in the spec here, but is in the reference implementation.)

Yes, but since it is serialized with Avro, it will always be an `int`:

```json
{
  "type": "int",
  "logicalType": "date"
}
```

Specifying this twice would lead to duplication of [the Avro spec](https://avro.apache.org/docs/1.12.0/specification/#date) into the Iceberg spec. The error in Iceberg-Rust did raise my eyebrow a bit since I would expect it to read the `int` without the `logicalType` as well because there is no ambiguity.

Reply:
I also encountered this before in [apache/iceberg-python#1208](https://github.com/apache/iceberg-python/pull/1208) and [apache/iceberg-python#1211](https://github.com/apache/iceberg-python/pull/1211). There's also this accompanying devlist thread https://lists.apache.org/thread/2gq7b54nvc9q6f1j08l9lnzgm5onkmx5